crates/bevy_tasks/src/task_pool.rs

```diff
@@ -105,89 +256,176 @@ impl TaskPool {
+    pub fn build() -> TaskPoolBuilder {
+        TaskPoolBuilder::new()
+    }
+
+    fn new_internal(builder: TaskPoolBuilder) -> Self {
         let (shutdown_tx, shutdown_rx) = async_channel::unbounded::<()>();
 
-        let executor = Arc::new(async_executor::Executor::new());
+        let compute_executor = Arc::new(async_executor::Executor::new());
+        let async_compute_executor = Arc::new(async_executor::Executor::new());
+        let io_executor = Arc::new(async_executor::Executor::new());
 
-        let num_threads = num_threads.unwrap_or_else(num_cpus::get);
+        let total_threads =
+            crate::logical_core_count().clamp(builder.min_total_threads, builder.max_total_threads);
+        tracing::trace!("Assigning {} cores to default task pools", total_threads);
 
-        let threads = (0..num_threads)
-            .map(|i| {
-                let ex = Arc::clone(&executor);
-                let shutdown_rx = shutdown_rx.clone();
+        let mut remaining_threads = total_threads;
 
-                // miri does not support setting thread names
-                // TODO: change back when https://github.com/rust-lang/miri/issues/1717 is fixed
-                #[cfg(not(miri))]
-                let mut thread_builder = {
-                    let thread_name = if let Some(thread_name) = thread_name {
-                        format!("{} ({})", thread_name, i)
-                    } else {
-                        format!("TaskPool ({})", i)
-                    };
-                    thread::Builder::new().name(thread_name)
-                };
-                #[cfg(miri)]
-                let mut thread_builder = {
-                    let _ = i;
-                    let _ = thread_name;
-                    thread::Builder::new()
-                };
+        // Determine the number of IO threads we will use
+        let io_threads = builder
+            .io
+            .get_number_of_threads(remaining_threads, total_threads);
 
-                if let Some(stack_size) = stack_size {
-                    thread_builder = thread_builder.stack_size(stack_size);
-                }
+        tracing::trace!("IO Threads: {}", io_threads);
+        remaining_threads = remaining_threads.saturating_sub(io_threads);
+
+        // Determine the number of async compute threads we will use
+        let async_compute_threads = builder
+            .async_compute
+            .get_number_of_threads(remaining_threads, total_threads);
 
-                thread_builder
+        tracing::trace!("Async Compute Threads: {}", async_compute_threads);
+        remaining_threads = remaining_threads.saturating_sub(async_compute_threads);
+
+        // Determine the number of compute threads we will use
+        // This is intentionally last so that an end user can specify 1.0 as the percent
+        let compute_threads = builder
+            .compute
+            .get_number_of_threads(remaining_threads, total_threads);
+
```

Comment:
should there be a trace! for the # of compute threads?

Reply:
Done.